File: c++/cote/GJM/연결된_부분집합의_개수.cpp
#include <iostream>
#include <random>
#define N 1000
#define TEST_CASES 10

using namespace std;

void dfs(int cur, vector<vector<int>>& graph, bool* visited){
	visited[cur] = true;

	for (int i : graph[cur]) {
		if (! visited[i]) {
			dfs(i, graph, visited);
		}
	}
}

int solve(vector<vector<int>>& graph) {
	int count = 0;
	bool visited [graph.size()];
	memset(visited, false, sizeof(visited));

	for (int i = 0; i < graph.size(); i++) {
		if (visited[i]) continue;
		dfs(i, graph, visited);
		count++;
	}

	return count;
}

int main(void) {
	default_random_engine gen(43);

	for (int it = 0; it < TEST_CASES; it++) {
		uint n = gen() % (N - 3) + 3;
		
		vector<vector<int>> graph {n, vector<int>()};

		for (int i = 0; i < n; i++) {
			for(int j = i + 1; j < n; j++) {
				if (gen() % 127 > 0)  continue;
				graph[i].push_back(j);
				graph[j].push_back(i);
			}
		}

		cout << n << " " <<solve(graph) << endl;
	}

	return 0;
}

File: go/cote/GJM/연결된_부분집합의_개수.go
package main

import (
	"fmt"
	"math/rand"
)

const (
	TEST_CASES = 10
	N          = 1000
)

func initialize(random *rand.Rand) (ret [][]int) {
	n := random.Intn(N)
	ret = make([][]int, n)

	for i, _ := range ret {
		for j := i + 1; j < n; j++ {
			if random.Float32() < 0.25 {
				ret[i] = append(ret[i], j)
			}
		}
	}
	return
}

func dfs(cur int, visited []bool, graph [][]int) {
	visited[cur] = true

	for _, next := range graph[cur] {
		if !visited[next] {
			dfs(next, visited, graph)
		}
	}

}

func solve(graph [][]int) (count int) {
	visited := make([]bool, len(graph))

	for i := 0; i < len(graph); i++ {
		if !visited[i] {
			dfs(i, visited, graph)
			count++
		}
	}

	return count
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		graph := initialize(random)
		fmt.Println(solve(graph))
	}
}
